avogadro/io/lammpsformat.cpp

```diff
@@ -391,24 +392,150 @@ bool LammpsFormat::read(std::istream& inStream, Core::Molecule& mol)
+                                   const Core::Molecule& mol)
 {
   return false;
 }
 
-std::vector<std::string> LammpsFormat::fileExtensions() const
+std::vector<std::string> LammpsTrajectoryFormat::fileExtensions() const
 {
   std::vector<std::string> ext;
   ext.push_back("dump");
   return ext;
 }
 
-std::vector<std::string> LammpsFormat::mimeTypes() const
+std::vector<std::string> LammpsTrajectoryFormat::mimeTypes() const
 {
   std::vector<std::string> mime;
   mime.push_back("text/lammps");
   return mime;
 }
 
+LammpsDataFormat::LammpsDataFormat() {}
+
+LammpsDataFormat::~LammpsDataFormat() {}
+
+bool LammpsDataFormat::read(std::istream& inStream, Core::Molecule& mol)
+{
+  return false;
+}
+
+bool LammpsDataFormat::write(std::ostream& outStream, const Core::Molecule& mol)
+{
+  // Title
+  if (mol.data("name").toString().length())
+    outStream << mol.data("name").toString() << std::endl;
+  else
+    outStream << "LAMMPS data file generated by Avogadro" << std::endl;
+
+  std::ostringstream massStream, atomStream, bondStream;
+  double xmin, xmax, ymin, ymax, zmin, zmax;
+
+  size_t numAtoms = mol.atomCount();
+  outStream << to_string(numAtoms) << " atoms\n";
+
+  size_t numBonds = mol.bondCount();
+  outStream << to_string(numBonds) << " bonds\n";
+
+  // A map of atomic symbols to their quantity.
+  size_t idx = 1;
+  Array<unsigned char> atomicNumbers = mol.atomicNumbers();
+  std::map<unsigned char, size_t> composition;
+  for (Array<unsigned char>::const_iterator it = atomicNumbers.begin(),
+                                            itEnd = atomicNumbers.end();
+       it != itEnd; ++it) {
+    if (composition.find(*it) == composition.end()) {
+      composition[*it] = idx++;
+    }
+  }
+
+  outStream << composition.size() << " atom types\n";
+
+  // Masses
+  massStream << "Masses\n\n";
+  std::map<unsigned char, size_t>::iterator iter = composition.begin();
+  while (iter != composition.end()) {
+    massStream << iter->second << "   " << Elements::mass(iter->first);
+    ++iter;
+  }
+  massStream << std::endl << std::endl << std::endl;
+
+  if (numAtoms) {
+    // Atomic coordinates
+    atomStream << "Atoms\n\n";
+    for (Index i = 0; i < numAtoms; ++i) {
+      Atom atom = mol.atom(i);
+      if (!atom.isValid()) {
+        appendError("Internal error: Atom invalid.");
+        return false;
+      }
+      Vector3 coords = atom.position3d();
+      if (i == 0) {
+        xmin = coords[0];
+        xmax = coords[0];
+        ymin = coords[1];
+        ymax = coords[1];
+        zmin = coords[2];
+        zmax = coords[2];
+      } else {
+        xmin = std::min(coords[0], xmin);
+        xmax = std::max(coords[0], xmax);
+        ymin = std::min(coords[1], ymin);
+        ymax = std::max(coords[1], ymax);
+        zmin = std::min(coords[2], zmin);
+        zmax = std::max(coords[2], zmax);
+      }
+
+      atomStream << std::setw(static_cast<size_t>(log(numAtoms)) + 1)
```

Comment:
Personally, I'd use strnprintf() or similar - I find most people can read the format strings more easily than the C++ designations